062-worker-pool/proposal.md

```diff
@@ -0,0 +1,157 @@
+# Summary
+
+This proposal outlines the support for worker pool. A worker pool is a group
+of Concourse workers, that can be associated with one or multiple teams.
+
+# Motivation
+
+One or multiple Concourse workers can be associated with a team by adding 
+`--team <team name>` option when launching `concourse worker`. Then all 
+pipeline containers of the team will be dispatched to the workers. The 
+relationship between workers and teams are N-to-1:
+
+```
+  Workers (0..n) <----------------> (0..1) Team
+```
+
+This is a useful mechanism to isolate some teams' workload or to provide
+dedicate workers to some teams.
+
+In practice, some big projects may split their pipelines to multiple small
+teams. When a project that has multiple teams requests dedicate workers,
+it is hard to give. Because the project actually wants some workers sharing
+with all of its teams, but Concourse only support workers per team.
+
+For example, a big project of my company has 100 teams. By evaluating
+workload of those teams, the project requests 20 dedicate workers. However, due
+to the restriction of 1 worker serving at most 1 team, we have to deploy at
+least 100 workers (1 worker for a team) to the project. Moreover, each team 
+having a single worker is unreliable, if the worker is down, then all builds of
+the team will fail. If we deploy 2 workers for each team for redundancy, then
+we need to deploy 200 workers. Comparing 200 workers we need to deploy with 20
+workers they request, it is a huge waste.
+
+# Proposal
+
+With introducing a new concept "worker pool", we build N-N relationship between
+workers and teams:
+
+```
+  Workers (0..n) <----> (1) Worker Pool (1) <------> (0..n) Teams
+```
+
+* A worker pool may contain 0 or multiple workers.
+* A worker can be added to at most 1 worker pool.
+* A worker may also not belong to any worker pool.
+* If a worker is added to a worker pool, then it cannot be associated with a
+team directly.
+* If a worker is not added to any worker pool, then it can still be associated
+with a team directly for backward-compatibility.
+* A worker pool may be assign to 0 or multiple teams.
+* A team can be associated with at most 1 worker pool.
```

Comment:
Interesting - do you think this is a necessary restriction, or is this just being conservative to start out?

It kind of seems like a team could be associated with multiple, and just union all the worker pools together to get one big set of candidates. The final selection would then be based on platform/tag configuration within the set of workers.

For example, maybe there's a pool of Windows workers and a pool of Darwin workers that teams can opt in to. I guess those could be in one big pool, but maybe there'd be some reason to make those platforms opt-in - maybe the operators are  being careful about sharing them because they don't containerize as effectively.

🤷‍♂️

Reply:
My original thoughts is, a worker pool is nothing more than a name, thus workers with different platforms, tags can join the same pool. Given a pool may attach to multiple teams, if a team can also attach to multiple pools, then the relationship between pool and team is N-to-N, which will make management more complicated.

But I can also imagine an use case like, teamA and teamB share a Darwin pool as Darwin workloads are limited, but each of them has a dedicate Linux pool.